app/WebParsing/ReqParser.hs

```diff
@@ -247,46 +288,107 @@ justParser = do
     markInfoParser = do
         Parsec.try (fmap Left percentParser <|> fmap Left letterParser <|> fmap Right infoParser)
 
--- parse for single course with or without cutoff OR a req within parentheses
-courseParser :: Parser Req
-courseParser = Parsec.between Parsec.spaces Parsec.spaces $ Parsec.choice $ map Parsec.try [
+-- Parses for single course or a group of programs with or without cutoff OR a req within parentheses
+-- Programs need to be parsed in groups because of the concatenation issue
+-- explained in the docstring of `programGroupParser`
+courseOrProgParser :: Parser Req
+courseOrProgParser = Parsec.between Parsec.spaces Parsec.spaces $ Parsec.choice $ map Parsec.try [
     parParser,
     cutoffParser,
     justParser,
+    programOrParser,
     rawTextParser
     ]
 
+programParser :: Parser Req
+programParser = do
+    Parsec.spaces
+    program <- Parsec.manyTill Parsec.anyChar $ Parsec.choice $ map (Parsec.try . Parsec.lookAhead) [
+        degreeType,
+        programSuffix,
+        progGroupSeparator,
+        progOrSeparator,
+        Parsec.oneOf ".;" >> return "",
+        Parsec.eof >> return ""
+        ]
+    return $ PROGRAM program
+
+-- Flattens nested ORs into a single OR
```

Comment:
This is a helper function. Move it to the bottom of the file (after all of the parsers).